dsgrid/dataset/growth_rates.py

```diff
@@ -0,0 +1,92 @@
+import logging
+
+import pyspark.sql.functions as F
+from pyspark.sql.types import IntegerType
+
+from dsgrid.dimension.base_models import DimensionType
+from dsgrid.exceptions import DSGInvalidQuery
+from dsgrid.query.models import ExponentialGrowthDatasetModel
+from dsgrid.utils.spark import get_unique_values
+
+
+logger = logging.getLogger(__name__)
+
+
+def apply_growth_rate_123(
+    dataset: ExponentialGrowthDatasetModel,
+    initial_value_df,
+    growth_rate_df,
+    time_columns,
+    model_year_column,
+    pivoted_columns,
+):
+    """Applies the growth rate to the initial_value dataframe.
+
+    Parameters
+    ----------
+    initial_value_df : pyspark.sql.DataFrame
+    growth_rate_df : pyspark.sql.DataFrame
+    time_columns : set[str]
+    model_year_column : str
+    pivoted_columns : set[str]
+
+    Returns
+    -------
+    pyspark.sql.DataFrame
+
+    """
+    initial_value_df, base_year = _check_model_years(
+        dataset, initial_value_df, growth_rate_df, model_year_column
+    )
+    gr_df = growth_rate_df
+    for column in pivoted_columns:
+        gr_col = column + "__gr"
+        gr_df = gr_df.withColumn(
+            gr_col,
+            F.pow((1 + F.col(column)), F.col(model_year_column).cast(IntegerType()) - base_year),
+        ).drop(column)
+
+    dim_columns = set(initial_value_df.columns) - pivoted_columns - time_columns
+    # TODO DT: data_source needs some thought. They are different in these two dfs.
+    # And this should be dimension_query_name instead of dimension type
+    # What is the data_source of the resulting df?
+    if DimensionType.DATA_SOURCE.value in dim_columns:
+        dim_columns.remove(DimensionType.DATA_SOURCE.value)
+    if DimensionType.DATA_SOURCE.value in gr_df.columns:
+        gr_df = gr_df.drop(DimensionType.DATA_SOURCE.value)
+
+    df = initial_value_df.join(gr_df, on=list(dim_columns))
```

Comment:
I think you (or I) could use the new formula functionality here

Reply:
Are you talking about DatasetExpressionHandler?

Reply:
The multiplication of value col and _gr col

Reply:
Yes, this PR doesn't have to address this